Add `endpoint` option to `Builder`

BenKanouse opened the pull request:
https://github.com/jrochkind/faster_s3_url/issues/9

Currently the `Builder`'s initializer accepts a `host` param which assumes https. I like to use minio for local development and it isn't ideal to run it with `https`.

The normal Aws::S3::Client gem accepts an `:endpoint` option to by-pass this issue.

Here is a section of the docs that explains this option:

```
:endpoint (String, URI::HTTPS, URI::HTTP) — Normally you should not configure the :endpoint option directly. This is normally constructed from the :region option. Configuring :endpoint is normally reserved for connecting to test or custom endpoints. The endpoint should be a URI formatted like:
'http://example.com'
'https://example.com'
'http://example.com:123'
```
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html


## Performance tests

<details>

<summary>public_bench -- this branch</summary>

``` 
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3     2.311k i/100ms
         FasterS3Url    32.292k i/100ms
Calculating -------------------------------------
          aws-sdk-s3     23.427k (± 0.9%) i/s   (42.69 μs/i) -    117.861k in   5.037506s
         FasterS3Url    325.072k (± 0.5%) i/s    (3.08 μs/i) -      1.647M in   5.068149s
                   with 95.0% confidence
```

</details>


<details>

<summary>presigned_bench -- this branch</summary>

``` 
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3   143.000 i/100ms
aws-sdk-s3 with custom headers
                       156.000 i/100ms
 re-used FasterS3Url     2.652k i/100ms
re-used FasterS3Url with cached signing keys
                         4.336k i/100ms
re-used FasterS3URL with custom headers
                         1.954k i/100ms
new FasterS3URL Builder each time
                         1.930k i/100ms
re-used WT::S3Signer     8.772k i/100ms
new WT::S3Signer each time
                         2.354k i/100ms
Calculating -------------------------------------
          aws-sdk-s3      1.582k (± 2.0%) i/s  (631.98 μs/i) -      7.865k in   5.002142s
aws-sdk-s3 with custom headers
                          1.480k (± 2.4%) i/s  (675.47 μs/i) -      7.488k in   5.100616s
 re-used FasterS3Url     25.842k (± 0.8%) i/s   (38.70 μs/i) -    129.948k in   5.032903s
re-used FasterS3Url with cached signing keys
                         39.025k (± 2.3%) i/s   (25.62 μs/i) -    195.120k in   5.030891s
re-used FasterS3URL with custom headers
                         19.703k (± 2.0%) i/s   (50.75 μs/i) -     99.654k in   5.087072s
new FasterS3URL Builder each time
                         21.853k (± 1.8%) i/s   (45.76 μs/i) -    110.010k in   5.059017s
re-used WT::S3Signer     89.431k (± 2.0%) i/s   (11.18 μs/i) -    447.372k in   5.025458s
new WT::S3Signer each time
                         26.529k (± 2.1%) i/s   (37.69 μs/i) -    131.824k in   5.006578s
                   with 95.0% confidence
```

</details>


<details>

<summary>public_bench -- on v1.1 of the project</summary>

``` 
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3     2.354k i/100ms
         FasterS3Url    42.537k i/100ms
Calculating -------------------------------------
          aws-sdk-s3     24.112k (± 0.8%) i/s   (41.47 μs/i) -    122.408k in   5.082080s
         FasterS3Url    443.088k (± 0.7%) i/s    (2.26 μs/i) -      2.254M in   5.092750s
                   with 95.0% confidence
```

</details>

<details>

<summary>presigned_bench.rb -- on v1.1 of the project</summary>

``` 
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3   149.000 i/100ms
aws-sdk-s3 with custom headers
                       137.000 i/100ms
 re-used FasterS3Url     2.429k i/100ms
re-used FasterS3Url with cached signing keys
                         4.155k i/100ms
re-used FasterS3URL with custom headers
                         2.154k i/100ms
new FasterS3URL Builder each time
                         2.299k i/100ms
re-used WT::S3Signer     8.976k i/100ms
new WT::S3Signer each time
                         2.419k i/100ms
Calculating -------------------------------------
          aws-sdk-s3      1.571k (± 2.0%) i/s  (636.63 μs/i) -      7.897k in   5.054203s
aws-sdk-s3 with custom headers
                          1.540k (± 1.5%) i/s  (649.54 μs/i) -      7.672k in   5.002016s
 re-used FasterS3Url     26.317k (± 0.9%) i/s   (38.00 μs/i) -    133.595k in   5.082447s
re-used FasterS3Url with cached signing keys
                         46.601k (± 2.0%) i/s   (21.46 μs/i) -    232.680k in   5.030958s
re-used FasterS3URL with custom headers
                         22.195k (± 1.4%) i/s   (45.06 μs/i) -    112.008k in   5.064084s
new FasterS3URL Builder each time
                         25.327k (± 1.1%) i/s   (39.48 μs/i) -    126.445k in   5.003145s
re-used WT::S3Signer    100.899k (± 1.2%) i/s    (9.91 μs/i) -    511.632k in   5.084384s
new WT::S3Signer each time
                         27.649k (± 0.7%) i/s   (36.17 μs/i) -    140.302k in   5.078609s
                   with 95.0% confidence
```

</details>

BenKanouse committed:
Add `endpoint` option to `Builder`

Currently the Builder's initializer accepts a host param
which assumes https. I like to use minio for local
development and it isn't ideal to run it with https.

The normal Aws::S3::Client gem accepts an :endpoint
option to by-pass this issue.

Here is a section of the docs that explains this option:

```
:endpoint (String, URI::HTTPS, URI::HTTP) — Normally you should not configure the :endpoint option directly. This is normally constructed from the :region option. Configuring :endpoint is normally reserved for connecting to test or custom endpoints. The endpoint should be a URI formatted like:
'http://example.com'
'https://example.com'
'http://example.com:123'
```
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html

In this commit I add the extra option into the builder class while
maintaining the current functionality for the host option.

BenKanouse committed:
Refactor uri logic out of the Builder

The builder class is pretty large and it has to do a lot of extra edges
cases around finding the correct host and endpoint. In this commit I
create a FasterS3Url::Uri class and move all uri specific logic to that
class.

The builder class caches it and all of it's attributes are cached, so it
shouldn't have a negative impact on performance.

jrochkind commented:
Thank you for this, and for the good description!

 I'm sorry I've been very busy in other areas last week and this, but this is on my radar. I will try to look at it today, if not, know that I haven't forgotten you, but if you don't hear from me for a bit, please feel free to ping again on this issue!

jrochkind commented:
OK, so `host` -- you are right that when I implemented/documented it I hadn't totally thought it ought and it didn't quite work in that way. 

But I _have_ ended up using it for when you have a CloudFront distribution (or any other CDN) on top of S3. For public unsigned urls, you just need to change the hostname to be your CDN hostname.  And this is a convenient tool to do it, in a speedy way (much speedier than AWS SDK), applying appropriate (a bit idiosyncratic) S3 escaping rules for keys, etc. 

So I think we do need to keep the `host` argument there, working as it did -- at least for unsigned URLs. Perhaps host and endpoint are mutually exclusive, and if you specify both you get an ArgumentException? Or we just doc which takes priority.

jrochkind commented:
OK thanks for this! 

Splitting it out into a separate class is nice and readable code organization wise. (I'd probably call the class BaseURI instead of just URI). 

BUT.  The whole point of this thing is performance. We need to make sure splitting it into a different class, or these changes in general, don't harm performance significanty -- especially in cases where this new feature is NOT being used. 

Would you be interested in trying to run the in-repo benchmarking stuff against this new one -- maybe find a way to run the old unchanged code against the new code -- to see how it effects performance? 

Thank you!

jrochkind commented:
Also add a CHANGELOG.md entry in PR please, with a URL to this PR!

BenKanouse commented:
Thanks for getting back to me! I can get back to this early next week.

It sounds like I have the following TODOs:
- [ ] Run the performance specs before and after this change.
- [ ] Update the CHANGELOG.md with a link to this PR.
- [ ] Rename the new class from URI to BaseURI. (Assuming performance is good with this class).

I am still a little confusing about the `host` feature. Does that mean when you use it you do specify the bucket name in the `host` string you provide? If so that makes a lot of sense, and I could update the specs to match that pattern.

jrochkind commented:
Is the difference between endpoint and host just that endpoint allows you to include a port too? I am having trouble understanding what's up here -- can you give me illustrative examples of what you would pass in as endpoint and where you would get it?

jrochkind commented:
Hey if I understand what endpoint is doing, I think this could and shoudl actually be a lot simpler. 

Endpoint is just to allow you to specify `http` instead of `https`, and a port in addition to a host?

This code can be much simpler.  If Endpoint is specified, just use the string given to construct the url, no?

Places in the code that have `"https://" + self.host + path` just need to be changed to have logic like:

    if self.endpoint
         self.endpoint + path
    else
        "https://" + self.host + path
     end

Even better, if on initialization a host is given and not an endpoint, just SET the endpoint to "https://#{host}", and then all subsequent logic can just be in terms of endpoint -- `host` is just a shortcut for setting `endpoint` that assumes `https` and no port. 

I think this can just be a few lines of change, and doesn't need a separate class, or instantiating a stdlib `URI` object -- things I was worried about their effects on performance. 

My use of `host`.  I have an S3 Bucket.  So something in the S3 bucket might be at the public URL "https://#my-bucket.s3.amazonaws.com/somepath/something.jpg".  No signature, it's a public URL.  

But i have an [AWS Cloudfront](https://aws.amazon.com/cloudfront/) CDN in front of it, so to access it through AWS Cloudfront I actually need to access it at `https://d987adfa07df.cloudfront.net/somepath/something.jpg`, so I set host to `d987adfa07df.cloudfront.net`.

BenKanouse commented:
I started down that expect path, and I ran into 2 problems.

1. When endpoint it an IP address AWS has to form the path differently by adding the bucket. Which I try and point in this [spec](https://github.com/jrochkind/faster_s3_url/pull/10/files#diff-55397705043290c57d8e318e16f642ff8250ababeaac2526eb5da8da66b0aee4R90). I might be able to make that spec clearer by adding a starts_with assertion. So the path changes as well as the host/endpoint.
2. The `canonical_headers` needs to not include the `http` or `https`. So when I have an `endpoint` param I need to be able to strip the `scheme` off of it.

Both the `path` and the `canonical_headers` are in the `string_to_sign` so it is important that those variables are exactly correct for the signature. So it makes it all it little more complicated than just changing the resulting url.

I do think you have a great idea about using the host to set an endpoint, which might simplify the logic! I didn't fully understand that host variable before, I can look into this next week. Thanks!

jrochkind commented:
Ah, i see what you mean about additional requirements. Hm, how about this @BenKanouse:

Right now it looks like the only specs you added are to your new URI object?

Can you add some actual public API tests -- write tests where you test all the pertinent different kinds of "endpoint" params (numeric IP, http vs https), and tests that the output is correct in all the ways you mention -- the canonical headers for correct signature (you could compare the signature against what the AWS-SDK does to make sure it is right, I think some existing tests may do that?), the extra path info for numeric IP etc. 

Write enough specs that if the specs pass, then we're good, regardless of inside implementation. Write specs for the behavior required, not the internal implementation. Presumably your current implementation would pass!

Then, I think I'll have some time this week to take a stab at writing the simplest possible implementation that passes your specs. Like explain to me everything it needs to do that you know about by writing the specs that will test it, and I'll see what i can do. 

Make sense?  Interested?

BenKanouse commented:
I did try to add those specs [here](https://github.com/jrochkind/faster_s3_url/pull/10/files#diff-55397705043290c57d8e318e16f642ff8250ababeaac2526eb5da8da66b0aee4R81) and [here](https://github.com/jrochkind/faster_s3_url/pull/10/files#diff-55397705043290c57d8e318e16f642ff8250ababeaac2526eb5da8da66b0aee4R133) in the `builder_spec.rb` file in my first commit.

My hope was that file is testing the public API, so I added those specs in the first commit before any of my refactors in the second.

It sounds like we discovered some more edge cases that I should add. 👍🏻

jrochkind commented:
OK even if we had the embedded BaseURI class, I would want those specs at the top level documenting the public API, right!

Can you add them, along with all the edge cases you can think of, such that if the tests pass, as far as you know the thing will be good?

I don't want to make you do a lot of other work spinning your wheels; if you can do this, I can try my idea for the simplest possible implementation that would pass specs!

BenKanouse committed:
Update changelog for endpoint feature

This feature was added to match the AWS endpoint param option documented
in AWS here:

https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html

BenKanouse commented:
Today I did the following:
1. Added extra specs to builder_spec.rb
2. Ran the performance tasks on version 1.1 and my branch and posted the results to the description of this PR.
3. Added to the Changelog.

I wasn't able to do `endpoint to "https://#{host}"` because the specs would start failing. It turns out that `host` dose not act like AWS's endpoint param. I made a spec here showing that they are not equal:
https://github.com/jrochkind/faster_s3_url/pull/10/files#diff-55397705043290c57d8e318e16f642ff8250ababeaac2526eb5da8da66b0aee4R80

The host param wants to make a url like this:
```
"https://my.example.com/some/directory/file.jpg"
```

While AWS's endpoint param makes a url like this:
```
"https://my-bucket.s3.amazonaws.com/some/directory/file.jpg"
```

While still much faster than AWS the performance did drop some after adding this feature.

I'm starting to think that maybe this project shouldn't support this feature. It feels like the point of this project is to create quick make URLs and if we add all the AWSs features we might end up with complex code that gets slowed down. For instance I needed to parse this logic in order to understand how AWS deals with endpoints: https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-s3/lib/aws-sdk-s3/endpoint_provider.rb#L30 Which makes me think there are more edge-cases out there.

jrochkind commented:
Hey @BenKanouse !

I have been pretty busy with other stuff recently!  I see what you mean. 

Yes, the host param is completely separate, my further suggestion may have been infeasible, just ignore current `host` entirely for now I think!

If you think you can capture the top-level  public API interacton _specs_ fully, such that an I am happy to spend some time doing the _implementation_ to match the specs. It might be possible and worthwhile to capture the majority of cases even if there are some weird edge cases not captured? I don't, for instance, think we need to error on all the conditions that the code you link to errors on, we can let the user try with "impossible" config. 

But if you don't think it seems feasible for you to do this, then that would be a signal that it's not feasible to implement right now!

I don't entirely understand what is going on myself though, I haven't had time to look more in depth into it!

If you think the specs you already have are sufficient for a useful implementation, let me know and I'll work on it!

BenKanouse commented:
That sounds correct to me!

I do have those specs in the `spec/builder_spec.rb` file on this branch.

jrochkind commented:
You think the specs in there now are sufficient to spec correct behavior? They include the header stuff you mentioned and anything else required?  If so, I'll take a look!

BenKanouse commented:
Yes, I think so. The specs are not asserting on what the params are in the url, so I am not directly asserting that the `canonical_headers` is correct. Instead I am asserting that it is the same url/params that AWS will create.

For example I'm doing this in many of the specs:

```
expect(builder.presigned_url(object_key)).to eq(aws_bucket.object(object_key).presigned_url(:get))
```

If we wanted to we could split the params off of the `builder.presigned_url(object_key)` and make more explicit assertions, but I'm not sure if that would be more or less confusing.

jrochkind commented:
OK, I'm sorry, I'm still getting confused. Why do you have a spec that says `does NOT match the AWS logic for the endpoint param`, why is the correct desired behavior to not match AWS logic in this case?

I haven't looked at this code in a while and at this point you probably have a fresher understanding of than me. Can you do me a favor and make it easier for me, can you help me identify exactly what specs we are talking about that specify the desired behavior at the top-level API? And then I would love to take a stab at writing an implementation to meet them.

jrochkind commented:
OK wait sorry should have rubber-ducked it, I think I figured it out thanks!

BenKanouse commented:
The ones that are `NOT match` are for the existing host behavior. Those specs might be pointing out that we are doing something with `host` that AWS's gem can't do.

The new behavior specs are under the `endpoint` sections.

jrochkind commented:
Closed for #14
